Proposal to release version 3.1.0.0 based on the attached changes.

abzalov opened the pull request:
Hello.

First of all, thanks for creating the repository!

I have done some work and propose to release version 3.1.0.0 based on the following changes:

**LockedList.cpp**
  3.1.0.0 - 19th April 2024
  * Added SilentCancel NSIS function to cancel async search (cancel by callback may not help - the callback is called too rarely and for found items only).
  * Added CleanupAsyncThread function to cleanup the state of async search thread before stating new search.
  * Added IsAsyncDone function to help determine the current async search thread state.
  * g_hThreadFiles handle now not closed after any async search to hold last search result (done, cancel or wait).
  * Increased array sizes for lists (files, modules, classes, captions, custom items and folders) from 128 to 256.

**SystemEnum.cpp**
  v2.0 - 19th April 2024
  * Fixed Wow64DisableWow64FsRedirection was not reverted to initial value on exit.
  * Re-wrote GetSystemHandleInformation() - now NtQuerySystemInformation used SystemExtendedHandleInformation (instead of obsolete SystemHandleInformation).
    Large number of handles led to overflow fields of SYSTEM_HANDLE structure (USHORT HandleValue and other).
  * Handle types are now cached to improve performance.
  * Added CheckFileDeviceType to avoid processing handle types that outside area of our interest (especially pipe handles):
     - NtQueryInformationFile or NtQueryObject may freeze
     - reduce the number of processed handles
  * Fixed erroneous ResetEvent(g_hFinishNow) in SystemFuncInit - be called several times during 1 search:
      (GetSystemHandlesCount and EnumSystemHandles) or (GetSystemModulesCount and EnumSystemProcesses).
    Now calling ResetEvent(g_hFinishNow) moved to SystemEnum - the main entry point of any unum, called only once
  * Now EnumSystemHandles does not handle handles of processes that could not be opened using OpenProcess in GetProcessFileName (System, System Idle, and other privileged ones).

Nadahar commented:
Hi @abzalov 

This was quite a lot of changes. C/C++ really isn't my thing, so it's hard for me to evaluate what all this does, so I figured I'd start by testing it. But, I get an error when trying to build. It seems to choke on `constexpr` - which seems to be introduced in C++11.

One of the things I really like with NSIS is its backwards compatibility, it doesn't dictate what systems you can make software for, unlike most tools these days. This plugin is also built with Visual Studio 2010. How much work is it to make this work without compromising compatibility?

abzalov commented:
Hi @Nadahar

C/C++ is also not my strong point :)
Therefore, there may be inconsistencies and shortcomings in my code, something may be done in an unconventional way.
But I think this is not critical and I will be corrected.

However, I tested the changes well - everything works fine and is backwards compatible.

As for constexpr, it can be replaced with #define:
```
//handle types cache
#define MAX_OBJECT_TYPE_NAME_SIZE 1024
#define MAX_CACHED_OBJECT_TYPE_NUMBER 257
static TCHAR g_ObjectTypeNames[MAX_CACHED_OBJECT_TYPE_NUMBER][MAX_OBJECT_TYPE_NAME_SIZE] = {0};
```

Backwards compatibility should not be a problem.

Another build error would be an undeclared FILE_DEVICE_CONSOLE constant.
I think it can be solved like this:
```
#ifndef FILE_DEVICE_CONSOLE
#define FILE_DEVICE_CONSOLE             0x00000050
#endif
```

Nadahar commented:
If you push the changes needed to make it build to your fork, this PR will be updated and I can try to test it.

abzalov commented:
I've done. I hope did it right in git.

Nadahar commented:
> I've done. I hope did it right in git.

It seems so - now it builds at least. I'll have to try to do some testing, but I must first wake up the part of my memory that revolves around this - it doesn't seem to be "up and running" at the moment 😉

Nadahar commented:
I've only tested it on Windows 10 this far, and it seems to work fine there. Do I observe a speed increase when using `LockedList::AddFile`? I remember that it used to be terribly slow - but my test now wasn't that bad.

Nadahar commented:
When testing on WinXP 64, it appear to no longer work. No error or anything, but it simply ignores the locked files and proceeds until the installer can't write the files. Can it be that some of the changes in a24fca6d293cdd05327348521eaf1dd1f5134007 breaks XP?

Nadahar commented:
On WinXP 32 it appears to hang at "25% Search" even when no files are actually locked.

edit: At my second attempt it hung at 26% - and it definitely hangs "hard" - "back" or "cancel" does nothing, I must use task manager to kill the process - and even it takes quite a while before it manages to kill it.

Nadahar commented:
It appears to be working well on Win7 32, so I don't think I need to test Win7 64 for now.

Nadahar commented:
I tried reverting a24fca6d293cdd05327348521eaf1dd1f5134007, but it made no difference with my XP tests. The two above problems still persist. Do you have any idea what changes can have caused this, as it's a bit too time-consuming to revert, build, build the installer and then test every commit on the VMs?

Typical suspects are changed API calls, changed struct definitions etc. The reason I suspected a24fca6d293cdd05327348521eaf1dd1f5134007 was because it changed `SystemHandleInformation`.

abzalov commented:
> I've only tested it on Windows 10 this far, and it seems to work fine there. Do I observe a speed increase when using `LockedList::AddFile`? I remember that it used to be terribly slow - but my test now wasn't that bad.

Yes, the speed should increase significantly. Because handles that are not related to files are discarded from consideration.
In particular, managed to avoid working with pipes.

abzalov commented:
I realized that there are problems with XP (32 and 64 bits).
I'll try to investigate and fix this.
Perhaps SystemExtendedHandleInformation is not supported there, maybe something else.

Nadahar commented:
According to my tests, as I stated above, `SystemExtendedHandleInformation` is not the culprit - at least not the only one as I experienced the same problems without this commit.

abzalov commented:
Fixed bug. Should be fine now.

Nadahar commented:
Now it works on WinXP 32, but the problem with WinXP 64 is still there. The behavior there isn't that it's hanging - it's simply that it behaves as if the plugin isn't installed at all. No search, no nothing - it just completely skippes everything related to LockedList.

WinXP 64 has shown itself to "behave differently" also in the past, AFAICR b42625c1278824311c0ea37a703414dca10626f6 was also to get WinXP 64 working. Problems with WinXP 64 also applies to Windows Server 2003.

What I remember vaguely is that on WinXP 64/Win2003, some of the API calls returns 0 for size even though there really is data to fetch. That's why b42625c1278824311c0ea37a703414dca10626f6 fixed it - we simply "ignore" the 0 size. Maybe the same must be done with the "new" logic too?

Nadahar commented:
There's something strange - I'm probably doing something wrong, but I'm unable to even trigger a breakpoint on one of the DLL "entry points" (`addFile`, `initDialog`) on WinXP 64. I've configured Visual Studio 2010 on the WinXP 64 VM, and I've used the very same method to debug the plugin on WinXP 64 in the past. But, now it's like the plugin "doesn't exist" at all when the installer is run on that VM.

abzalov commented:
To hook into the debugged LockedList I use calls CallInstDLL instead of direct calls to plugin LockedList:: ...

Because in the case of plugin calls LockedList:: ... - installer loads and uses the extracted library from $PLUGINSDIR, not the debugable one.

Example: CallInstDLL C:\open\LockedList\Plugins\x86-unicode-debug\LockedList.dll SilentSearch

Nadahar commented:
That might make it easier - do you mean that CallInstDLL let you load DLL's "dynamically"?

Anyway, what I do have worked in the past, I build the whole installer with the debug DLL - so that what it extracts from `$PLUGINSDIR` is the debuggable one. But, no luck this time - it's as if there's something wrong with the "loading" of the DLL itself or something. Sadly, NSIS isn't great on debugging, I don't really know how to get insight into that.

Nadahar commented:
Ok, there something really weird going on. I now built the last commit before your changes, and it still doesn't "load" (or do anything) on WinXP 64. At the same time, I know that it used to work because I explicitly tested it on this VM. I'm starting to think it's something wrong in my NSIS script - but what can that be - considering that it works on 32 bit AND on 64 bit Win10.

Btw, I've tried CallInstDLL now and it doesn't make a difference.

abzalov commented:
>I'm starting to think it's something wrong in my NSIS script - but what can that be - considering that it works on 32 bit AND on 64 bit Win10.

First of all, check the !addplugindir commands in the NSIS script. Check that the specified directory contains the newly built version of LockedList and not some other version.

abzalov commented:
I rechecked on Windows Server 2003 x64 and Windows x64 - everything seems to be working.
To check, I use LockedList::AddFolder, and the installer finds running applications from the specified directory.

You need to check the following points:
   - the LockedList plugin is actually loaded by your installer
   - the LockedList plugin file is exactly the one you used when compiling your installer - check the version or size of the file in bytes
All this can be done, for example, using Process Explorer from Sysinternals. True, you will have to look for an old version that will run on Windows XP.

Typically, when the NSIS installer starts using the plugin, it extracts it into the %TEMP% directory in the ns[XXXXX].tmp folder.
When the search dialog appears or after it, the plugin has started its work, and you can already search for this directory in TEMP or see Process Explorer - to understand which libraries are loaded by the installer process.

Nadahar commented:
The problem is weirer than that. It's not just that the debugging doesn't work - the plugin doesn't run either - it doesn't search for locked files, it doesn't handle locked files. It's as if the plugin doesn't exist. I've now downgraded to an older version of NSIS itself (to the version I think I was when this worked), but that made no change either. I'm sure this is something really stupid, but at the moment I'm all out of ideas of what can cause this.

I've also verified that another plugin I use works on the WinXP 64 installer - and it does - so it's not that it doesn't load plugins at all.

But, when thinking about it, there is one "strange" thing that is different with the loading of LockedList - all the other plugsin use their 32 bit versions since the NSIS installer I use is 32 bits - but for LockedList there is some "hack" with (also) loading the 64 bit DLL to be able to search 64 bit processes. Maybe this is where something goes wrong - to be honest I don't quite understand how this is supposed to work. I wouldn't think that a 32 bit process could load a 64 bit DLL at all.

Nadahar commented:
> Typically, when the NSIS installer starts using the plugin, it extracts it into the %TEMP% directory in the ns[XXXXX].tmp folder.
When the search dialog appears or after it, the plugin has started its work, and you can already search for this directory in TEMP or see Process Explorer - to understand which libraries are loaded by the installer process.

As I've said, the search dialog never appears - I don't think the plugin loads at all - and if you have tested this on Win2003 it's pretty clear that this is unrelated to your changes and might be something I will have to figure out independently from this PR.

abzalov commented:
>That might make it easier - do you mean that CallInstDLL let you load DLL's "dynamically"?

Yes everything is correct.

**Using the plugin using the CallInstDLL command.**
  - When the installer compiles: NSIS does NOT include the plugin library in the installer.
  - When the installer is running: the library is searched and loaded dynamically along the specified path from the local host file where the installer is running.

**Using the plugin using the LockedList::... command**
  - When the installer compiles: NSIS packages the plugin library into the installer. At the same time, it takes it from the directory specified by the !addplugindir command.
  - When the installer is running: the plugin library is first unloaded into the %TEMP% directory and then also loaded from there dynamically.

Nadahar commented:
This is from the LockedList readme, the about "magic stuff":

> As of v3.0.0.0, 64-bit module enumeration (for ::AddModule) is now
  supported. Simply extract a copy of LockedList64.dll to $PLUGINSDIR
  on 64-bit machines (see the LockedListKernel32.nsi example script).
  The same LockedList64.dll is used for both ANSI and Unicode
  installers.

Looking in `LockedListKernel32.nsi`, it is done like this, exactly like it's done in my script:
```nsis
  ${If} ${RunningX64}
    File /oname=$PLUGINSDIR\LockedList64.dll `${NSISDIR}\Plugins\LockedList64.dll`
  ${EndIf}
```
Maybe there is something wrong with my NSIS script in doing this - but confusingly, it works fine on Win10.

I have no idea how this can work at all - the file is called `LockedList64.dll`, not `LockedList.dll` as normal, so I would think there would be a reference to loading this somewhere in the plugin for this "magic" to work - but if so I haven't found the code.

abzalov commented:
By the way, yes, maybe the problem is in the LockedList64.dll library
If the installer must use it, then before compile your installer - need to make sure that you compile LockedList64.dll (from new sources) and place it in the directory ${NSISDIR}\Plugins.

Nadahar commented:
I have compiled and included debug versions of both the 32 bit and 64 bit versions - but as I said, the plugin never seems to load at all - not even the "32 bits version".

Nadahar commented:
The "magic" seems to happen here:

https://github.com/DigitalMediaServer/LockedList/blob/819e9e9bb08022fc75a6ffd9f90930002747e05d/Contrib/LockedList3/LockedList.cpp#L496-L505

abzalov commented:
Have you compiled both libraries - LockedList.dll for Win32, and LockedList64.dll for Win64?
To build the LockedList64.dll library - compile the project with configuration: Release64 and Platform x64

Nadahar commented:
Yeah, I have compiled both - but can't use `Release64` because I need the debug version.

abzalov commented:
> The "magic" seems to happen here:

The "magic" of using Win64 library by Win32 installer is using rundll32.exe. Look in the code "rundll32".
Win32 installer launches Win64 rundll32.exe, which in turn loads the Win64 library LockedList64.dll. That is, a separate process is launched.

Then rundll32.exe reports about found files to the installer that launched it.

Nadahar commented:
> The "magic" of using Win64 library by Win32 installer is using rundll32.exe. Look in the code "rundll32".
Win32 installer launches Win64 rundll32.exe, which in turn loads the Win64 library LockedList64.dll. That is, a separate process is launched.

My guess is that this is where something goes wrong - but I have no idea how to try to debug/figure it out.

abzalov commented:
Can you replace all LockedList:: calls with CallInstDLL commands passing arguments via Push?

For example:
`LockedList::AddFolder $INSTDIR`

This should turn to:
```
Push $INSTDIR
CallInstDLL C:\open\LockedList\Plugins\x86-unicode-debug\LockedList.dll AddFolder
```

And

```
LockedList::Dialog /autonext /heading $(LockedListDialogHeading) \
  /colheadings $(LockedListDialogAppHeader) $(LockedListDialogProcessHeader) \
  /menuitems $(LockedListDialogCloseMenuItem)  $(LockedListDialogCopyMenuItem) \
  /autoclose $(LockedListDialogCloseText) $(LockedListDialogKillText) $(LockedListDialogFailedText) $(LockedListCloseAll) \
  /searching $(LockedListSearching)
```

This should turn to:
```
Push "/autonext"
Push "/heading $(LockedListDialogHeading)"
Push "/colheadings $(LockedListDialogAppHeader) $(LockedListDialogProcessHeader)"
Push "/menuitems $(LockedListDialogCloseMenuItem)  $(LockedListDialogCopyMenuItem)"
Push "/autoclose $(LockedListDialogCloseText) $(LockedListDialogKillText) $(LockedListDialogFailedText) $(LockedListCloseAll)"
Push "/searching $(LockedListSearching)"
CallInstDLL C:\open\LockedList\Plugins\x86-unicode-debug\LockedList.dll Dialog
```

Nadahar commented:
I did try that earlier - with no difference. But, I must have mixed up some file versions or something (there's a lot of manual file copying and building going on here), because now at did it all again from scratch, and now I have the plugin running AND the debugger working on the latest commit before this PR.

So, now I will have to switch commit and do it all again and see if I can get it to work 😉

Nadahar commented:
I've done it all again with the version from this PR, and it the plugin doesn't work now - although I think I can see a very brief "flash" of what is the "search dialog". My debugger is attaching now, and I can follow the execution, so I'll have to try to figure out what goes wrong. My eyes need a little break at the moment though.

I followed it through `ShowDialog` and everthing seemed normal there - it looped until `g_fDone` was true. So, I guess I will have to attach to `FilesThread` to see more of what's going on.

Nadahar commented:
This is hard to debug. I clearly don't know enough about working with the Windows API to properly follow the logic here, especially the post/peek/translate/dispatch message stuff. So, I don't really know how to tell the difference between what happens and what ought to happen.

However, I just realized one thing. In my test, I'm locking a 64 bit DLL which I expect LockedList to pick up. That means that the "rundll32 trick" must work for my test to succeed. It could be that a lock on some other file which isn't a "64 bit module" would work. This could explain why your test works while mine fails. I don't know of a simple way to lock a different file for locking on XP though, since the method I normally use relies on PowerShell, but I guess it could be possible to dig up some VB script that does it. The way I currently lock the file is by letting the application I'm "installing" run, and it will use 64 bit DLLs on a 64 bit system.

What I have found this far, and that I think is a bit strange (although that might be wrong), is that when debugging through this code:
https://github.com/DigitalMediaServer/LockedList/blob/819e9e9bb08022fc75a6ffd9f90930002747e05d/Contrib/LockedList3/LockedList.cpp#L2720-L2736
..it only iterates the loop once. After the first "wait" it proceeds, `PeekMessage` doesn't find any message and the next wait the returns `dwWaitRes = 0` which results in `break`. As such, it never seems to retrieve any information from "LockedList64.dll".

Nadahar commented:
I think I've gotten closer now. I've built and tested each commit, and it turns out that it's 098e73ce1ac4d133044f336585c6dbc63a4f2f3f that breaks it. That commit doesn't do that much, so maybe you have an idea of why this breaks it?

Nadahar commented:
I just made one more experiment to confirm: I used all the commits except 098e73ce1ac4d133044f336585c6dbc63a4f2f3f (I actually just reversed 098e73ce1ac4d133044f336585c6dbc63a4f2f3f) and built the plugin, and then it works. So, 098e73ce1ac4d133044f336585c6dbc63a4f2f3f is definitely the culprit. It's strange how this doesn't prevent it working on more recent 64 bit systems, but something has probably changed in Windows in the meanwhile.

abzalov commented:
Interesting.
In fact, CreateEvent should be created as a manual reset (2nd param should be TRUE):
`  g_hFinishNow = CreateEvent(NULL, TRUE, FALSE, NULL);`

Can you enable 098e73c commit and modify this code to check?

Nadahar commented:
> Can you enable https://github.com/DigitalMediaServer/LockedList/commit/098e73ce1ac4d133044f336585c6dbc63a4f2f3f commit and modify this code to check?

Sure.

Nadahar commented:
I didn't help.

abzalov commented:
I think that the installer script that reproduces the problem will help us.

Nadahar commented:
That's easy, it's this one: https://github.com/DigitalMediaServer/DigitalMediaServer/blob/master/src/main/external-resources/nsis/setup.nsi

That said, most of what's happening there isn't related to LockedList.

Nadahar commented:
I don't think the installer script itself is important - as long as it copies `LockedList64.dll` into `$PLUGINSDIR` before calling `Dialog`. But, the test needs to lock a 64 bit "module" that is in use - I guess that is harder to "fake".

This might be helpful or it might be a red herring, but when I debugged this, it *seemed* to me like the 64 bit "child process" did something the first time it was called but not the second. It is first called with `GetSystemModulesCount64`, and then with `EnumSystemProcesses64`. The first time I could "see" that it was doing something because there was a slight delay and the screen "flashed", the second time I could see nothing. But, that could also be because of some caching mechanism that made the second call go much quicker.

My thought was just that maybe somehow, the "event" is already signaled when the child process is spawned. But, that shouldn't be possible, should it? I mean, these are separate processes with nothing in common..?

The problem is that I'd need to debug the "child process" spawned via `rundll32`, but since I have to attach the debugger manually, I'd need to make it wait until I can get attached before it starts processing. With the installer process itself, that is easy because of the GUI - I attach to it while it waits for me to press "next" on one of the screens. But, the "child process" does its thing and closes much faster than I can attach a debugger. I can't really think of a way to make it "wait" until I "permit it" to continue - can you?

abzalov commented:
The installer process represented by LockedList.dll (32-bit) and the launched process rundll32.exe (64-bit) work on a request-response principle.
The request and its parameters are specified on the command line to launch rundll32.exe.
rundll32.exe sends responses via WM_COPYDATA messages sent to the installer window.

The LockedList.dll plugin receives and processes received messages in the CallbackWindow64WndProc procedure. Both types of responses are processed there.

GetSystemModulesCount64 - requests the number of 64-bit modules (here the answer will be only 1).
EnumSystemProcesses64 - asks to list all 64-bit modules (there should be >= 0 answers here).

The entry point functions in LockedList64.dll are GetSystemModulesCount64W and EnumSystemProcesses64W.


>I can't really think of a way to make it "wait" until I "permit it" to continue - can you?

Typically, in order for the debugger to attach to a process, you need to make it wait. MessageBox (or sleep as a last resort) will help with this. Insert it at the beginning of the function (GetSystemModulesCount64W and EnumSystemProcesses64W) and then you can take your time connecting the debugger.

Nadahar commented:
Can I really call a message box from a DLL? In my world a DLL has no GUI, so I didn't really consider it. If that works, it should be simple enough.

abzalov commented:
The ability to use the GUI depends only on the process.
If the process is a console process, then MessageBox cannot be used, in other cases it is possible.

rundll32.exe is a GUI application, not a console application.

Nadahar commented:
The messagebox trick works.

Nadahar commented:
I see one strange thing right away. The rundll32 process seems to hang somehow, it doesn't disappear from the process list even after the installer is closed. The debugger can also stay attached to it, although with only assembly view :(

Nadahar commented:
I think I've got it now 💡 

When using rundll32 and calling a function directly, `StartEnumeratingNow()` is never called and `g_hFinishNow` is never created. Thus, when it comes to `WaitForSingleObject` some error is returned which makes it abort. What I don't understand now is why it seemingly worked on Win10 - but maybe something was off with my testing. It makes sense that it would refuse to wait for a non-existing event... and the check is only if the result is `WAIT_TIMEOUT` or not - so the result could be an error, and we wouldn't pick up on that. `LastError` isn't checked either - so it would "look like" an ordinary abort.

edit: I just confirmed by testing that inserting a `StartEnumeratingNow()` call in `EnumSystemProcesses64W` makes it work.

Nadahar commented:
@abzalov I pushed a new commit that I think fixes it: 2babdab08edeb14f2b8eb3e325fac411e4b3b5c9. I've done a preliminary test already, but I'll do a full test using multiple OSes with release versions now.

The question is if `StartEnumeratingNow()` should be called from any of the other "externally called" (rundll32) functions, but I don't think `g_hFinishNow` is used by the other(s)?

That said, I'm still confused about how `g_hFinishNow` is supposed to work when using rundll32. Isn't the idea that if the user aborts, this event is set, which leads to the loops aborting? I can't quite see how `g_hFinishNow` in the 64 bit subprocess will be set?

Maybe 
https://github.com/DigitalMediaServer/LockedList/blob/819e9e9bb08022fc75a6ffd9f90930002747e05d/Contrib/LockedList3/LockedList.cpp#L2733
...should be modified to also wait for `g_hFinishNow` - and if triggered and the process is still running, somehow terminate the process and leave?

abzalov commented:
>I think I've got it now 💡

Exactly!
You are well done!

So it was my mistake, I thought that cancel event needed and only works on the 32-bit side.

To simply fix this the way it worked before my changes you need to call StartEnumeratingNow(); in EnumSystemProcesses64W and GetSystemModulesCount64W after
`if (IsWindow(hCallerWindow) && EnableDebugPriv())`

But if we simply create a cancel event on the x64 side, then the events will be independent. The cancellation signal from a 32-bit library will not reach a 64-bit library.
This is how it worked before - the cancel signal to g_hFinishNow did not reach the rundll32 process - the installer stopped but rundll32 continued to work until it completed the job.

Apparently need to somehow pass the original event handle to the 64-bit library and use it there.
I think this is possible through pass handle via command line args and get event listening through DuplicateHandle.

I understood the problem.
I can fix this a little later and commit it.

Thanks for your investigation!

Nadahar commented:
Actually, since it has gotten so much faster now - I don't know if it matters if the 64-bit process continues its task in the background, as long as the installer doesn't have to wait for it. It's up to you - how much trouble is it worth to "forward" the signal?

Nadahar commented:
Since I got the impression that you might not be so familiar with Git, you should do a `git pull` before you write new code or make commits, so that you get the commit I made copied to your local repository. Otherwise, there will be problems when you try to push your changes later.

abzalov commented:
> @abzalov I pushed a new commit that I think fixes it: 2babdab

Commit 2babdab is good.
But you also need to add a call to StartEnumeratingNow(); in GetSystemModulesCount64W - there the g_hFinishNow event is also used internally.

abzalov commented:
> Actually, since it has gotten so much faster now...

In general, yes, but it's kind of wasteful. I'll see how quickly I can do this.

Nadahar commented:
> But you also need to add a call to StartEnumeratingNow(); in GetSystemModulesCount64W - there the g_hFinishNow event is also used internally.

Are you sure? I can't find that it does, and also `GetSystemModulesCount64W` returns a result as it is now. I guess it wouldn't hurt to create it anyway, I just didn't see a reason to.

abzalov commented:
> Are you sure? I can't find that it does, and also `GetSystemModulesCount64W` returns a result as it is now. I guess it wouldn't hurt to create it anyway, I just didn't see a reason to.

I checked, indeed, the event is not used inside yet. But that may change one day.

abzalov commented:
To be honest, I don’t really understand why the LockedList::AddModule function is needed when there is a LockedList::AddFile.

AddFile will always find the application that loaded the library or any other executable file using an open handle to the loaded module in this process.
Thus AddFile is enough and does not need to run a 64-bit process to check all handles in the OS.

Nadahar commented:
I don't know - I just know that `AddModule` is much, much faster. It's when you use `AddFile` that it gets really slow. I haven't been using `AddFile` because it's so slow - maybe that's the issue, or maybe there are some other situations where `AddFile` won't pick it up? Or maybe they simply created `AddModule` first.

Edit: After reading way too many old posts, I found [a comment that concerns `AddFile` and `AddModule`](https://forums.winamp.com/forum/developer-center/nsis-discussion/250053-lockedlist-plug-in?p=3752752#post3752752):
> it will only work on files that have a handle open on them (DLL's do not when they are locked). And yes using AddFile on DLL will not work.

Nadahar commented:
@abzalov It sounds difficult to signal the 64 bit process to me, so I was thinking of a simpler, although a bit cruder, way to stop it. My idea is implemented in 83c5eb1f72fdf04308f25a0753455e129ed90a38. Please note that this doesn't actually work, so I must have made some mistake, but I guess it should be sufficient to see what I was thinking of.

Nadahar commented:
I think I have fixed the bug I had made in 72446ba08b11fe0ec6609617940bb1e55cffba3c (I didn't understand that `MsgWaitForMultipleObjects` added its own "internal" event), so now it seems to work.

It's still questionable whether this is a good approach, and it's hard to test properly, but it seems to work.

Nadahar commented:
@abzalov I have now done more thorough testing of my suggested "fix", and as far as I can tell it works like it should. So, I pushed it here. If you don't like it, let me know and I'll remove the commit again.

Nadahar commented:
@abzalov Are you still there?

abzalov commented:
Hello @Nadahar, I'm here.

I don’t know how to release this into the Test branch.
Perhaps I cloned your repository, but needed to make changes to mine. I definitely need to learn how to work with Git :)

The changes should be as follows. But first need to revert your changes.
[LockedList.patch](https://github.com/DigitalMediaServer/LockedList/files/15144148/LockedList.patch)

abzalov commented:
If necessary, I can explain the changes and why your approach will not work as intended. But it's tomorrow.

Nadahar commented:
I'll handle the git stuff - that I know 😉 

I don't necessarily need an explanation, I know that my solution was "kind of a hack" - I just didn't want to spend too much time on that problem (this only applies if the user cancels the search).

Nadahar commented:
I do have some problems with the .patch file though, Git claims: `error: No valid patches in input`. I'll see if I can figure it out, I can see that there are changes in the file but something makes Git unable to "understand" them.

Nadahar commented:
@abzalov I have removed the commit from this PR, but I'm not sure if I can get something useful out of the patch file. The `patch` command also refuse to use it with: `patch: **** Only garbage was found in the patch input.`.

If you've only made changes to `LockedList.cpp` it might be easier if you just attach the updated file to a comment here - and I can use that to update this PR.

Nadahar commented:
I give up on the `.patch` file. Something must have gone wrong when you created it - because as far as I can tell, it doesn't actually include any changes. Removed lines are supposed to start with a `-` and new lines with a `+` (modified lines have both "remove" and "add"). There are no actual changes in the file as far as I can tell.

abzalov commented:
I'm sorry, I actually copied the Git Desktop output. This was probably wrong.

I attached changed files.
[LockedList3.zip](https://github.com/DigitalMediaServer/LockedList/files/15148553/LockedList3.zip)

Nadahar commented:
Does 1f8f73387cfba91c728e2e17b2035929b821426d look correct? I haven't studied it yet, but it does compile at least.

abzalov commented:
Yes, it's okay

Nadahar commented:
If you want to get your local git repo "up to date" with this branch, so that you can make changes and push (without getting errors/conflicts), do (assuming that your fork is called `origin`):
```bash
git fetch origin
git reset --hard origin/master
```
Please be aware that the "hard reset" will overwrite all files in your working tree, so don't do that if you have already made some changes.

abzalov commented:
>I know that my solution was "kind of a hack"

In fact, your solution was not bad, but not in the case of rundll32.

When testing, it turned out that 2 rundll32 processes are always launched. Can be seen in Process Explorer from Sysinternals.
It first runs the 32-bit version of rundll32 from C:\Windows\SysWOW64\, then the process sees that LockedList64.dll is a 64-bit library and so then runs the 64-bit version of rundll32 from C:\Windows\System32\ as a child process.

As a result, TerminateProcess kills only the 32-bit rundll32 process, which actually does nothing but waits for the 64-bit rundll32 process it spawned to finish its work.

This behavior is independent of the process launch option - ShellExecuteEx or CreateProcess. I checked both options, the behavior is the same.
Also, this behavior does not depend on whether just rundll32.exe or the full path to the 64-bit version of C:\Windows\System32\rundll32.exe is specified.

For testing, I made a long pause (Sleep) in the x64 library in the EnumSystemProcesses64W function so that I could cancel the search.
You can cancel the search by moving from the LockedList page to the previous installer page or simply closing the installer window, or by using LockedList::SilentCancel.

Nadahar commented:
Ah, thanks for the investigation. I didn't even consider that such a thing could happen, but I shouldn't be surprised.

I know how to cancel the search, I just found it hard to test because it all goes too fast for me to be able to study it or attach a debugger. But yeah, a long sleep would solve that.

Nadahar commented:
I think I understand most of the changes. I guess I could be worried about "sharing" an object between processes, but since an "event" already is a "synchronization object", it's probably fine. It's a bit clever to share it like that, it avoids the whole problem of trying to "contact" the other process to signal that it's time to wrap up when a cancel action occurs 👍 

I will do my "regular round" of testing later, and report back.

Nadahar commented:
All my tests were successful, so I think we're there 👍 

There might be some modifications necessary in the "changelog notes", I'm not sure. What I would like to do is to rebase the "fixes" into their original commits, if it's OK with you. It makes the Git history "cleaner" and easier to understand if somebody wants to try and figure something out in the future.

When that's done, it might be easier to compare the "changelog notes" with the commits and see if any updates are necessary.

Nadahar commented:
@abzalov I'm done with my rebasing. I made a backup before I rebased it in [DigitalMediaServer/LockedList/Backup](https://github.com/DigitalMediaServer/LockedList/tree/backup).

I haven't changed anything apart from some slight tweaks to some of the commit messages and some additions to 87b8f7d8822e4ab456e67aed06ab5c958626feb9 (updated the readme++). Apart from that, I've only shuffled changes around to where they "should" originally have been made.

I'm not sure if everything is mentioned as it should in the "changelog" now - if you want to make modifications, just do another fetch/reset, make your changes and push a new commit. I can rebase the changes in "where they belong" later.

If you don't think any further changes are needed, let me know and I will merge it.

abzalov commented:
Everything is fine. Thank you!

Nadahar commented:
I'll tag and make a release.

Nadahar commented:
@abzalov Released: https://github.com/DigitalMediaServer/LockedList/releases/tag/v3.1.0.0

abzalov commented:
Thanks for your assistance!

abzalov commented:
I don't know how difficult it is.
But I would also add a description of the changes from SystemEnum.cpp to the release description:
  * Fixed Wow64DisableWow64FsRedirection was not reverted to initial value on exit.
  * Re-wrote GetSystemHandleInformation() - now NtQuerySystemInformation used SystemExtendedHandleInformation (instead of obsolete SystemHandleInformation).
    Large number of handles led to overflow fields of SYSTEM_HANDLE structure (USHORT HandleValue and other).
  * Handle types are now cached to improve performance.
  * Fixed erroneous ResetEvent(g_hFinishNow) in SystemFuncInit - be called several times during 1 search:
      (GetSystemHandlesCount and EnumSystemHandles) or (GetSystemModulesCount and EnumSystemProcesses).
    Now calling ResetEvent(g_hFinishNow) moved to SystemEnum - the main entry point of any enum, called only once
  * Now EnumSystemHandles does not handle handles of processes that could not be opened using OpenProcess in GetProcessFileName (System, System Idle, and other privileged ones).

Nadahar commented:
That's not difficult - but is it correct? I've added what was in the commit messages - and that should cover changes in both files..?

I thought that some of those changes were reversed too - so I'm a bit confused. But, if you are sure that those changes are "real" and aren't already mentioned, I can add it.

Nadahar commented:
@abzalov I can confirm that I am confused. I thought that I used the changes from the commit messages - but apparently that's not what I did. I must have used those from the "changelog" in the readme - which again I must have copied from LockedList.cpp so that the changes in SystemEnum.cpp isn't included.

Ideally, we should update the readme to also have all the changes together with the release text. Can you check what needs to be modified now that I've used the commit messages for the release?

abzalov commented:
>But, if you are sure that those changes are "real" and aren't already mentioned, I can add it.
Yes, all changes are included in the commit, I double-checked.

Now everything is fine in the release description.
Just some small clarifications (they might not be worth the effort):
  - The following point is a whole it is better to combine it:
  Re-wrote GetSystemHandleInformation() - now NtQuerySystemInformation() uses SystemExtendedHandleInformation (instead of obsolete SystemHandleInformation).
    A large number of handles led to overflow fields of SYSTEM_HANDLE structure (USHORT HandleValue and others).
  - Only 1 point left unmentioned
  Increased array sizes for lists (files, modules, classes, captions, custom items and folders) from 128 to 256.


In Readme.txt, in the Change log section for item 3.1.0.0 - 30th April 2024, there are not enough items from SystemEnum.cpp - the rest is correct:
  * Fixed Wow64DisableWow64FsRedirection was not reverted to initial value on exit.
  * Re-wrote GetSystemHandleInformation() - now NtQuerySystemInformation used SystemExtendedHandleInformation (instead of obsolete SystemHandleInformation).
    Large number of handles led to overflow fields of SYSTEM_HANDLE structure (USHORT HandleValue and other).
  * Handle types are now cached to improve performance.
  * Fixed erroneous ResetEvent(g_hFinishNow) in SystemFuncInit - be called several times during 1 search:
      (GetSystemHandlesCount and EnumSystemHandles) or (GetSystemModulesCount and EnumSystemProcesses).
    Now calling ResetEvent(g_hFinishNow) moved to SystemEnum - the main entry point of any enum, called only once
  * Now EnumSystemHandles does not handle handles of processes that could not be opened using OpenProcess in GetProcessFileName (System, System Idle, and other privileged ones).


In Readme.txt, in the More information section, it might be worth adding a description of the new functionality for users:
  - Added SilentCancel NSIS function to cancel async search (cancel by callback may not help - the callback is called too rarely and for found items only).
The rest of the changes are bug fixes, performance improvements and added reliability. Therefore, it seems to me that they should not be described in this section. But maybe I'm wrong.

Thank you!

abzalov commented:
Description of the LockedList::SilentCancel function is already in the Readme.txt

Nadahar commented:
@abzalov I've made the edits to the release.

Regarding `Readme.txt` I was a bit unsure if your last commit meant that I shouldn't add the stuff about `SilentCancel`? Is this what you meant the changes should be? b412e33f5b1cfe1c1263a72990fdf44ff540ebe1

abzalov commented:
I think everything is great now.
Thanks for your participation!

Nadahar commented:
@abzalov I haven't pushed b412e33f5b1cfe1c1263a72990fdf44ff540ebe1 to master yet, but I assume you think that commit is good then? I'm not doing a new release just for the documentation update though, but it will at least be on Github... and will be included in a future release.

abzalov commented:
Yes, everything is fine.
You know better how to do this better, I still don’t quite understand such subtleties.

PS. I started learning git, hope it will become a little clearer :) http://www-cs-students.stanford.edu/~blynn/gitmagic/intl/en/pr01.html

Nadahar commented:
> PS. I started learning git, hope it will become a little clearer :) http://www-cs-students.stanford.edu/~blynn/gitmagic/intl/en/pr01.html

If you're on Windows, I'd just like to add that I love [TortoiseGit](https://tortoisegit.org/). While some people frown upon using GUI tools, I think there are some things that they do so much better than the command line, like merging conflicts, showing overviews, logs, commits etc. I don't use TortoiseGit exclusively, I use it in combination with the command line. It can be mixed as you want anyway, because all TortoiseGit does is translate your GUI actions into command line commands that it runs (and it shows you the commands so you can learn). Things like rebasing, reset, switching branch etc I typically do from the command line. But I commit using the GUI (because if shows me exactly what I'm committing in a very clear way), and I do all kind of log/history viewing, diff viewing, blame etc from the GUI. These things are just personal preference, and everybody will have to figure out what works for them, but I certainly think using something like TortoiseGit makes the whole thing a little less confusing in that it gives you a better overview.

I've only tried GitHub desktop briefly many moons ago, and I didn't find it very appealing. I don't remember exactly why, but I generally don't like to be too tied to GitHub's functionality anyway. I want to do (and learn/get used to) general Git, which will work anywhere.

Edit: I took a look at the tutorial you linked to, and I must say that while it contains many "correct" things, I question some of the things I saw. Like the comparison of `pull` and `push` - these are "opposite" functions, and it's not a case of using one or the other. I also think it jumps into writing commands way too quickly, and should spend more time explaining the concepts. Also, the repeated use of `git init` is confusing to me - this is something you almost never use (it's just used when you create a NEW git repository when starting a new project). So, I get the feeling that the author hasn't really "gotten the grasp" of Git as a tool. That said, Git is very versatile, so different people undoubtedly use it for different things and in different ways.

I don't have what I'd consider a "good" Git tutorial at hand, but I did find that this explains many of the concepts that I find essential to understand: https://git-scm.com/docs/user-manual

I'd also say that I think the most important thing to understand is the structure of Git itself - that it consists of commits, that all have their SHA1 commit identifier, which always "builds" on a parent commit (except the very first one). Once a commit is made, it is immutable - it can never be changed. Branches and tags are merely "pointers" to different SHA1 commits, which makes it easier for humans to navigate. You can make any number of branches or tags point to the same commit (which would be pointless, but conceptually they can). Git is essentially a "blockchain". You can "modify" a commit (amend, rebase), but what you're in fact doing then is making a new commit that is similar to the old one, but has a new SHA1 hash and represents a "branch" in the git tree. All of the commits that follow must then also be modified (although their content can be identical), because every commit is always "hard linked" to its parent. In my view, understand these concepts is what "unlocks" Git for you - until these things are clear, it's all "mysterious" and difficult to understand. Once you understand the basic concepts, everything becomes easy and logical. That doesn't mean that you know every command or parameter - but that's not necessary. You understand the concept, and can look up ways to do what you want when you need it.

I'd also like to add that, while this is personal preference and not everybody agrees, I've learned to really dislike "merge commits". I avoid them if at all possible (and it is always possible unless you work with somebody that insists on using them). When first introduced to Git, I got the impression that merging was "an essential part" - and I think that's slightly misleading. A "merge commit" differs from regular commits in that it has more than one parent commit (usually two, but it can be more). This causes great difficulty when trying to figure out what happened later, because the commit has "two or more separate diffs" with two or more parents. One merge commit isn't a big hurdle to overcome, but when you get many of them, things can quickly become unmanageable in my opinion. What makes them even worse is that IF you have a "conflict" when making a merge commit, the changes you make to solve the conflict becomes kind of "invisible" - so if you try to rebase/make changes to some commit later and you have to traverse a merge commit, you have to do the "conflict resolution" manually over and over again. So, instead of merging I use rebasing, which means to "adjust" a set of commits (a branch for example) to fit with another set of commits (that you want to "merge" with) by resolving the conflicts by modifying the commits themselves. That way, there are no "lost conflict resolution" like there are with merge commits. You also get a much cleaner and more understandable, linear history for when you need to go back to figure out how something came to be at a later stage.

As a consequence, I almost never use `git pull` and `git merge` (they are basically the same, `git pull` is `git fetch`+ `git merge`). IF I have to use them, I use the `--ff-only` argument, which makes sure that no merge commit is created (it then fails if a merge is necessary instead).

Many people are very happy with merge commits and will very much disagree with me in this, but it is my impression that these people usually don't spend much time tracing what happened, "investigating" if you want. They will instead try to "fix" a problem without going back under figuring out where something went wrong in the first place, often causing new "issues" and making everything more complicated. But, by working this way, they never really experience the "downside" of using merge commits - and I see how it can be more convenient to use if you never plan on looking back anyway.

Nadahar commented:
@abzalov I just did a sizeable edit to my post and see that you have seen it while I was editing, so I make this comment just to alert you of the edit (I will delete this comment, it's just used to trigger a notification).

abzalov commented:
Thank you! Very valuable!

Nadahar commented:
@abzalov I took the job to "merge" the "old" `Readme.txt` with the "GitHub standard" `README.md`. In addition, I split the changelog part out to `CHANGELOG.md`. This means that it will be presented as the "default documentation" on GitHub (as opposed to people having to explicitly open `Readme.txt`), which makes the documentation changes "more relevant" even if a new release hasn't been made. It also, in my opinion, makes the documentation more accessible in general:

https://github.com/DigitalMediaServer/LockedList/blob/master/README.md
https://github.com/DigitalMediaServer/LockedList/blob/master/CHANGELOG.md

abzalov commented:
Yes, that would be great!
